### PR TITLE
[CPU] Unify distribution tile size selection heuristic.

### PR DESCRIPTION
**Pull request by hanhanW**
The revision removes special distribution heuristics for CPUDefault. It does not impact lit tests because the unified method covers some of them. The revision introduces a DistributionHeuristicConfig struct to hold heuristic factors. This avoids having four optional function arguments. We want to land two changes together because the intermediate stage looks bad.

The removal changes are in https://github.com/openxla/iree/pull/15928/commits/13d4b8a90368f5049c7623a803b3070ef6829c84 The DistributionHeuristicConfig struct changes are in https://github.com/openxla/iree/pull/15928/commits/723560774155d1ed8c32e96978691623ca1da59f.

### CODE REVIEW

**Comment by hanhanW** [comment]
The review would be easier if you review https://github.com/openxla/iree/commit/13d4b8a90368f5049c7623a803b3070ef6829c84 first, then https://github.com/openxla/iree/commit/723560774155d1ed8c32e96978691623ca1da59f

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 1]
```diff
@@ -1320,8 +1335,9 @@ static LogicalResult setRootConfig(func::FuncOp entryPointFn,
 static LogicalResult setRootConfig(func::FuncOp entryPointFn,
                                    IREE::LinalgExt::FftOp fftOp) {
   assert(!getLoweringConfig(fftOp) && "expected lowering_config is not set");
+  DistributionHeuristicConfig config;
   SmallVector<int64_t> distTileSizes =
-      getDefaultDistributedLevelTileSizes(fftOp);
+      getDefaultDistributedLevelTileSizes(fftOp, config);
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Can be `getDefaultDistributedLevelTileSizes(fftOp, DistributionHeuristicConfig{});`?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 2]
```diff
@@ -1409,19 +1425,15 @@ setDefaultGenericOpRootConfig(func::FuncOp entryPointFn,
         DispatchLoweringPassPipeline::CPUDefault);
   }
 
-  SmallVector<int64_t> minTileSizes = getMinTilingSizesForEachDim(
+  DistributionHeuristicConfig config;
```
Comment (pzread) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I feel like `distConfig` might be a clearer name given there can be other configs

**Review by pzread — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: good point, thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: SGTM, thanks!

**Pull request merged**